### ISSUE TITLE
Use handle from component for asset animation (?)

### ISSUE THREAD

**Issue by alexfoxgill** [djeedai/bevy_tweening#101]
`MaterialMesh2dBundle<T>` has a `material: Handle<T>` component, and this can be used to animate without needing to pass the handle to the animator. I implemented this in my own project and it seems to work although perhaps there are some implications I am not aware of?

Code:

```rust
#[derive(Component)]
pub struct AssetComponentAnimator<T: Asset> {
    pub state: AnimatorState,
    tweenable: BoxedTweenable<T>,
    speed: f32,
}

impl<T: Asset> AssetComponentAnimator<T> {
    pub fn new(tweenable: impl Tweenable<T> + 'static) -> Self {
        Self {
            state: AnimatorState::Playing,
            tweenable: Box::new(tweenable),
            speed: 1.0,
        }
    }

    pub fn speed(&self) -> f32 {
        self.speed
    }

    pub fn tweenable_mut(&mut self) -> &mut dyn Tweenable<T> {
        self.tweenable.as_mut()
    }
}

pub struct AssetComponentTarget<'a, T: Asset> {
    assets: ResMut<'a, Assets<T>>,
    pub handle: Handle<T>,
}

impl<'a, T: Asset> AssetComponentTarget<'a, T> {
    pub fn new(assets: ResMut<'a, Assets<T>>) -> Self {
        Self {
            assets,
            handle: Handle::weak(HandleId::default::<T>()),
        }
    }

    pub fn is_valid(&self) -> bool {
        self.assets.contains(&self.handle)
    }
}

impl<'a, T: Asset> Targetable<T> for AssetComponentTarget<'a, T> {
    fn target_mut(&mut self) -> &mut T {
        self.assets.get_mut(&self.handle).unwrap()
    }
}

pub fn asset_component_animator_system<T: Asset>(
    time: Res<Time>,
    assets: ResMut<Assets<T>>,
    mut query: Query<(Entity, &Handle<T>, &mut AssetComponentAnimator<T>)>,
    events: ResMut<Events<TweenCompleted>>,
) {
    let mut events: Mut<Events<TweenCompleted>> = events.into();
    let mut target = AssetComponentTarget::new(assets);
    for (entity, handle, mut animator) in query.iter_mut() {
        if animator.state != AnimatorState::Paused {
            target.handle = handle.clone();
            let speed = animator.speed();
            animator.tweenable_mut().tick(
                time.delta().mul_f32(speed),
                &mut target,
                entity,
                &mut events,
            );
        }
    }
}
```

**Comment by djeedai** [comment]
Hi,

You mean that instead of passing the `Handle<T>` into `AssetAnimator::new()`, you directly retrieve it via the `Query<>` inside the animator system, right?

Interesting. On one hand I think this makes sense, and makes it easier to create the `AssetAnimator` without having to have the `Handle<T>` already readily available from the same place creating the animator. It also probably makes things more Bevy-like. On the other hand, this limits the animator to only animate the component located on the same `Entity`. This is a restriction compared to the current variant, however this is also a common Bevy pattern. So... I guess yes it makes sense?

**Comment by alexfoxgill** [comment]
yes in the same way that an `Animator<T>` looks for a `T: Component` on the entity, the `AssetAnimator<T>` could look for a `Handle<T>: Component` on the entity. it seems consistent to me but i have not worked with larger or 3d projects

**Comment by djeedai** [comment]
Changed on `main`, this will be available with the next version when Bevy 0.12 is out in the next few days or so. Thanks for the suggestion!
